fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/FoldConstantRuleOnFE.java

```diff
@@ -0,0 +1,310 @@
+import org.apache.doris.nereids.trees.expressions.EqualTo;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.ExpressionEvaluator;
+import org.apache.doris.nereids.trees.expressions.GreaterThan;
+import org.apache.doris.nereids.trees.expressions.GreaterThanEqual;
+import org.apache.doris.nereids.trees.expressions.InPredicate;
+import org.apache.doris.nereids.trees.expressions.IsNull;
+import org.apache.doris.nereids.trees.expressions.LessThan;
+import org.apache.doris.nereids.trees.expressions.LessThanEqual;
+import org.apache.doris.nereids.trees.expressions.Like;
+import org.apache.doris.nereids.trees.expressions.Not;
+import org.apache.doris.nereids.trees.expressions.NullSafeEqual;
+import org.apache.doris.nereids.trees.expressions.Or;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.expressions.TimestampArithmetic;
+import org.apache.doris.nereids.trees.expressions.WhenClause;
+import org.apache.doris.nereids.trees.expressions.functions.BoundFunction;
+import org.apache.doris.nereids.trees.expressions.functions.PropagateNullable;
+import org.apache.doris.nereids.trees.expressions.literal.BooleanLiteral;
+import org.apache.doris.nereids.trees.expressions.literal.Literal;
+import org.apache.doris.nereids.util.ExpressionUtils;
+
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * evaluate an expression on fe.
+ */
+public class FoldConstantRuleOnFE extends AbstractExpressionRewriteRule {
+    public static final FoldConstantRuleOnFE INSTANCE = new FoldConstantRuleOnFE();
+
+    @Override
+    public Expression rewrite(Expression expr, ExpressionRewriteContext ctx) {
+        return process(expr, ctx);
+    }
+
+    @Override
+    public Expression visit(Expression expr, ExpressionRewriteContext context) {
+        return expr;
+    }
+
+    /**
+     * process constant expression.
+     */
+    public Expression process(Expression expr, ExpressionRewriteContext ctx) {
+        if (expr instanceof PropagateNullable) {
+            List<Expression> children = expr.children()
+                    .stream()
+                    .map(child -> process(child, ctx))
+                    .collect(Collectors.toList());
+
+            if (ExpressionUtils.hasNullLiteral(children)) {
+                return Literal.of(null);
+            }
+
+            if (!ExpressionUtils.isAllLiteral(children)) {
+                return expr.withChildren(children);
+            }
+            return expr.withChildren(children).accept(this, ctx);
+        } else {
+            return expr.accept(this, ctx);
+        }
+    }
+
+    @Override
+    public Expression visitEqualTo(EqualTo equalTo, ExpressionRewriteContext context) {
+        return BooleanLiteral.of(((Literal) equalTo.left()).compareTo((Literal) equalTo.right()) == 0);
```

Comment:
where to ensure all children are literal?

Reply:
process method

Reply:
process method only handled the expression implement `PropagateNullable`. is it enough?